src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -0,0 +1,330 @@
+    {
+        status = Status::InvalidInState;
+        ChipLogError(Zcl, "Microwave Oven Control: Operational State instance does not exist");
+        goto exit;
+    }
+
+    opState = gOperationalStateInstance->GetCurrentOperationalState();
+    if (opState == to_underlying(OperationalStateEnum::kStopped))
+    {
+        uint8_t reqCookMode     = 0;
+        uint32_t reqCookTime    = 0;
+        uint8_t reqPowerSetting = 0;
+        if (CookMode.HasValue())
+        {
+            // TODO: set Microwave Oven cooking mode by CookMode.Value().
+        }
+        else
+        {
+            // TODO: set Microwave Oven cooking mode to normal mode.
+        }
+
+        if (CookTime.HasValue())
+        {
+            reqCookTime = CookTime.Value();
+        }
+        else
+        {
+            // set Microwave Oven cooking time to 30 seconds(default).
+            reqCookTime = MicrowaveOvenControl::kDefaultCookTime;
+        }
+
+        if (PowerSetting.HasValue())
+        {
+            reqPowerSetting = PowerSetting.Value();
+        }
+        else
+        {
+            // set Microwave Oven cooking power to max power(default).
+            reqPowerSetting = mDelegate->GetMaxPower();
+        }
+        status = mDelegate->HandleSetCookingParametersCallback(reqCookMode, reqCookTime, reqPowerSetting);
+        goto exit;
+    }
+    else
+    {
+        status = Status::InvalidInState;
+        goto exit;
+    }
+
+exit:
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+void Instance::HandleAddMoreTime(HandlerContext & ctx, const Commands::AddMoreTime::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "MicrowaveOvenControl:  HandleAddMoreTime");
+    Status status;
+    uint8_t opState;
+
+    // Get Operational State instance
+    gOperationalStateInstance = GetOPInstance(mEndpointId);
+    if (gOperationalStateInstance == nullptr)
+    {
+        status = Status::InvalidInState;
+        ChipLogError(Zcl, "Microwave Oven Control: Operational State instance does not exist");
+        goto exit;
+    }
+
+    opState = gOperationalStateInstance->GetCurrentOperationalState();
+    if (opState == to_underlying(OperationalStateEnum::kStopped) || opState == to_underlying(OperationalStateEnum::kRunning) ||
+        opState == to_underlying(OperationalStateEnum::kPaused))
+    {
+        uint32_t finalCookTime;
+
+        finalCookTime = GetCookTime() + req.timeToAdd;
+        // if the added cooking time is greater than the max cooking time, the cooking time stay unchanged.
+        if (finalCookTime < kMaxCookTime)
+        {
+            status = mDelegate->HandleSetCookTimeCallback(finalCookTime);
+            goto exit;
+        }
+        else
+        {
+            ChipLogError(Zcl, "Failed to set cookTime, cookTime value is out of range");
+            status = Status::ConstraintError;
+            goto exit;
+        }
+    }
+    else // operational state is in error
```

Comment:
That's not what the test above tests.  Is there a reason we're not just testing for "state is error" like the spec says?